### PR TITLE
HPCC-13058 Check frunssh failure in run_thor

### PR DESCRIPTION
**Pull request by Michael-Gardner**
This now allows us to see when frunssh fails and exit out of the init script gracefully.

Signed-off-by: Michael Gardner Michael.Gardner@lexisnexis.com

### CODE REVIEW

**Comment by Michael-Gardner** [comment]
@xwang2713 Please take a look when you get the chance.

**Comment by Michael-Gardner** [comment]
which line are you referring to exactly?  I don't think I understand the question.

**Comment by Michael-Gardner** [comment]
in 531 it catches if the script we ran on the remote machine returned an error.  So I don't really want to throw an exception for it, because it's not frunssh that failed.

**Comment by xwang2713** [comment]
OK

**Comment by Michael-Gardner** [comment]
Ming you were right.  I went through some more test cases and figured out that I did need to catch it in the else as well.

**Comment by Michael-Gardner** [comment]
@richardkchapman ready for you to take a look at it.

**Comment by richardkchapman** [comment]
I see no issues but best if @jakesmith looks too...

**Comment by richardkchapman** [comment]
Should this target 5.4.0 ?

File: common/remote/rmtssh.cpp [hunk 1]
```diff
@@ -521,9 +523,22 @@ class CFRunSSH: public CInterface, implements IFRunSSH
...
+                        errCode = reply.item(n);
+                        errMsg.append(res.str());
+                    }
+                }
                 else
                     PROGLOG("%d: %s(%d):\n---------------------------\n%s\n===========================",n+1,slaves.item(n),reply.item(n),res.str());
+                    if (reply.item(n)){
+                        errCode = reply.item(n);
+                        errMsg.append(res.str());
+                    }
```
Comment (jakesmith) [inline common/remote/rmtssh.cpp]: The above 'if (reply.item(n))' block looks like it was supposed to be part of the else @ line 533....
it's indented, but block bracing is missing.

File: common/remote/rmtssh.cpp [hunk 2]
```diff
@@ -521,9 +523,22 @@ class CFRunSSH: public CInterface, implements IFRunSSH
...
+            }
+            if (errCode) {
```
Comment (jakesmith) [inline common/remote/rmtssh.cpp]: Style: No need for curlies.. we normally exclude for single lines.

File: initfiles/componentfiles/thor/run_thor [hunk 3]
```diff
@@ -54,6 +54,14 @@ while [ 1 ]; do
     else
         nslaves=`cat $instancedir/uslaves.start | wc -l`
         $deploydir/frunssh $instancedir/uslaves.start "/bin/sh -c '$deploydir/start_slaves %a $THORMASTER $THORMASTERPORT $logdir $instancedir $deploydir $THORNAME $PATH_PRE $logredirect'" -i:$SSHidentityfile -u:$SSHusername -pe:$SSHpassword -t:$SSHtimeout -a:$SSHretries -n:$nslaves 2>&1
+        FRUNSSH_RC=$?
+        if [[ ${FRUNSSH_RC} -gt 0 ]]; then
+          echo "frunssh could not connect to all slave machines"
+          echo "check your key and ssh settings across the cluster"
```
Comment (jakesmith) [inline initfiles/componentfiles/thor/run_thor]: Does the error code necessarily mean 'could not connect' ? 
Wouldn't frunssh now return the same code if, e.g. the script it was executing failed?

File: common/remote/rmtssh.cpp [hunk 4]
```diff
@@ -521,9 +523,22 @@ class CFRunSSH: public CInterface, implements IFRunSSH
                 if (res.length()==0)
                     PROGLOG("%d: %s(%d): [OK]",n+1,slaves.item(n),reply.item(n));
                 else if (strchr(res.str(),'\n')==NULL)
+                {
                     PROGLOG("%d: %s(%d): %s",n+1,slaves.item(n),reply.item(n),res.str());
+                    if (reply.item(n)){
+                        errCode = reply.item(n);
```
Comment (jakesmith) [inline common/remote/rmtssh.cpp]: Trivial: inconsistent brace formatting in method/changes. 
My rule is, if small keep it consistent with exiting formatting, if larger change, reformat at least the method you are altering to our preferred brace formatting conventions, i.e. brace on newline.

**Comment by jakesmith** [comment]
@Michael-Gardner - a few comments.

**Comment by Michael-Gardner** [comment]
@jakesmith I should have hit all the issues.

**Comment by Michael-Gardner** [comment]
@jakesmith modified with multiException, and changed the brackets to match the style in the rest of the file.

**Comment by jakesmith** [comment]
Looks good to merge to me.
@richardkchapman

**Comment by Michael-Gardner** [comment]
re-target to 5.4

**Pull request closed without merging**